src/GafferScene/MergeObjects.cpp

```diff
@@ -0,0 +1,1448 @@
+		{
+			hasDescendantPruned = true;
+		}
+	}
+
+	bool needsOriginalBoundMerge = false;
+	IECore::MurmurHash childBoundHash;
+	if( sourcePaths && !( mergeLocationData && mergeLocationData->m_childNamesData->readable().size() ) )
+	{
+	}
+	else if( hasDescendantPruned )
+	{
+		childBoundHash = outPlug()->childBoundsPlug()->hash();
+	}
+	else if( hasDescendantDestinations )
+	{
+		childBoundHash = outPlug()->childBoundsPlug()->hash();
+
+		if( inputExists && !sourcePaths )
+		{
+			needsOriginalBoundMerge = true;
+		}
+	}
+	else
+	{
+		// If this location is neither a destination itself, or the parent of a destination, then
+		// it must exists in the input.
+		assert( inputExists );
+
+		// Since we're not adding or removing anything, we can just pass through the input.
+		childBoundHash = inPlug()->boundPlug()->hash();
+	}
+
+	if( !( sourcePaths || needsOriginalBoundMerge ) )
+	{
+		h = childBoundHash;
+		return;
+	}
+
+	FilteredSceneProcessor::hashBound( path, context, parent, h );
+
+	h.append( childBoundHash );
+	if( needsOriginalBoundMerge )
+	{
+		inPlug()->boundPlug()->hash( h );
+	}
+
+	if( !sourcePaths )
+	{
+		return;
+	}
+
+	h.append( outPlug()->fullTransformHash( path ) );
+
+	const ThreadState &threadState = ThreadState::current();
+	tbb::task_group_context taskGroupContext( tbb::task_group_context::isolated );
+
+	const ScenePlug *effectiveSource = effectiveSourcePlug();
+
+	const IECore::MurmurHash reduction = tbb::parallel_deterministic_reduce(
+		tbb::blocked_range<size_t>( 0, sourcePaths->size() ),
+		IECore::MurmurHash(),
+		[&] ( const tbb::blocked_range<size_t> &range, const MurmurHash &hash ) {
+
+			ScenePlug::PathScope pathScope( threadState );
+			IECore::MurmurHash result = hash;
+			for( size_t i = range.begin(); i != range.end(); ++i )
+			{
+				pathScope.setPath( &((*sourcePaths)[i]) );
+				effectiveSource->boundPlug()->hash( result );
+				result.append( effectiveSource->fullTransformHash( (*sourcePaths)[i] ) );
+			}
+			return result;
+
+		},
+		[] ( const MurmurHash &x, const MurmurHash &y ) {
+
+			MurmurHash result = x;
+			result.append( y );
+			return result;
+		},
+		tbb::simple_partitioner(),
+		taskGroupContext
+	);
+
+	h.append( reduction );
+}
+
+Imath::Box3f MergeObjects::computeBound( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent ) const
+{
+	ConstObjectPtr parentHolder;
+	const SourcePaths *sourcePaths = findSources( mergeLocationPlug(), path, context, parentHolder );
+
+	// Find the bound from children of this location, either in the existing scene, or added
+	// by merges targetting child locations as destinations.
+
+	// To know if we need to check the child bounds, we need to know if there are any destinations below this
+	// location. We can efficiently query this from the mergeLocationData.
+	ConstMergeLocationDataPtr mergeLocationData = IECore::runTimeCast< const MergeLocationData >( mergeLocationPlug()->getValue() );
```

Comment:
I'm sure it does have some sort of performance implications. In a simplistic view of read performance, grabbing mergeLocations plug from the cache is one map lookup ( of an admittedly large map ), whereas finding the location in branchesPlug() requires a map lookup for each level of the hierarchy, and then you'd have to do another plug evaluation of the filter to check if anything is pruned ( which is also unique at each location ). But there probably are cases where the hash pressure is a much bigger problem than doing some map lookups. Probably could be reasonable to add some more code paths to check the tree data and the filter, and skip access to mergeLocationPlug when not needed.

Reply:
> In a simplistic view of read performance, grabbing mergeLocations plug from the cache is one map lookup

Not sure I've read this right, but it seems like you might be implying that most `__mergeLocation` evaluations are cache hits? But it looks like `MergeLocationsData::hash()` includes `path` in the hash, meaning that we're computing something at `__mergeLocation` every time, even when there's no merging going on at that location. That's what led me to think this might be a fair bit worse than doing an initial lookup on the TreeData (which _is_ cached).

I suggest we park this temporarily while focussing on getting a final `MeshAlgo::merge()` implementation sorted. But I do think it'd be worth doing a bit of further investigation before we release.

Reply:
I probably should have documented my thoughts more before moving on, took me a bit to remember this, but anyway:

Yes, we hit MergeLocations for every location, including locations where no merging is going on. But the work performed in locations where there is no merging happening is the same work we would need to perform anyway in order to check if we should access MergeLocations.

So the trade-off is just whether we cache this work between different locations accessing MergeLocations - caching it is a minor speedup, but consumes more cache.


I've added two new tests in [5d5ccc3](https://github.com/GafferHQ/gaffer/pull/5984/commits/5d5ccc37b53cfdd55b0c45107475df94ff334948)

`testPerformanceUnprocessedLocations` tests the case where we aren't processing most locations, and MergeLocationData is usually NullObject anyway, which gives us the best case of seeing a benefit from not bothering to access it.

`testPerformanceDeepPaths` tests the case where there are deep paths being accessed repeatedly - the children of a location will access MergeLocationData in order to check their parent, so there are lots of accesses to every location, and querying the tree is expensive, giving maximum benefit to caching MergeLocationData.

[19f5fdb](https://github.com/GafferHQ/gaffer/pull/5984/commits/19f5fdb8a5483e28a10ae2a5762f281537a740ba) adds some extra code to perform the extra tree tree query and skip accessing MergeLocationsData whenever possible.

This causes testPerformanceDeepPaths to get 10% slower, and testPerformanceUnprocessedLocations to get 5% faster.

There is probably an argument though that testPerformanceUnprocessedLocations is more plausible to occur in production scenes, and while the performance benefit isn't huge, it does reduce cache usage by 50% in this extreme case - so maybe it's worth it?

If you want to go with 5d5ccc3, I should clean it up tomorrow. If you don't think it's needed, it would be fine to just drop the last commit.

Reply:
I hadn't anticipated the need for the filter checks in `accessMergeLocation()` - with my naive BranchCreator-primed hat on, I think I assumed the tree had all the info we need. But it doesn't, because that only stores destinations, right, and things that need pruning aren't stored?

For what it's worth, I'm seeing a 10% improvement in `testPerformanceUnprocessedLocations` and no real change in `testPerformanceDeepPaths` here on my workstation (32 core Xeon). My general feeling is that the merging-a-small-number-of-locations-in-a-large-scene is most likely our most important use case, but we also have to balance that against the maintainability of the code. I'll leave it to your judgement...

Reply:
Alternative thought : is it possible to tighten up the hash for `mergeLocationPlug()`, so that we do fewer computes on that plug by omitting `path` whenever we can?